### PR TITLE
Implemented functionality and testing for limiting concurrent batches

### PR DESCRIPTION
**Pull request by plasne**
You can now give Batcher a WithMaxConcurrentBatches option that will limit the number of batches that are in inflight at any given time.

### CODE REVIEW

File: batcher.go [hunk 1]
```diff
@@ -317,44 +365,48 @@ func (r *Batcher) Start() (err error) {
...
 	flush := func(watcher IWatcher, batch []IOperation) {
-		if len(batch) > 0 {
-			if r.emitBatch {
-				r.emit(BatchEvent, len(batch), "", batch)
+		if len(batch) == 0 {
+			return
+		}
+		if r.emitBatch {
+			r.emit(BatchEvent, len(batch), "", batch)
+		}
```
Comment (awkwardindustries) [inline batcher.go]: Nice!

File: batcher.go [hunk 2]
```diff
@@ -392,9 +444,16 @@ func (r *Batcher) Start() (err error) {
 			case <-auditTimer.C:
 				// ensure that if the buffer is empty and everything should have been flushed, that target is set to 0
 				if len(r.buffer) < 1 && time.Since(lastFlushWithRecords) > r.maxOperationTime {
-					if r.trySetTargetToZero() {
+					eventRaised := false
```
Comment (awkwardindustries) [inline batcher.go]: Very minor, but I wonder if `failEventRaised` might be more clear.

File: README.md [hunk 3]
```diff
@@ -39,13 +39,15 @@ Some other terms will be used throughout...
...
+- __Inflight__: The number of batches that are currently being processed (i.e. the batch has been raised to the callback function on the Watcher, but has not not completed that function yet).
+
 ![topology](./images/topology.png)
 
 ## Features
 
 - __Datastore Agnostic__: Batcher does not process the Operations it batches, it just notifies the caller when a batch is ready for processing. This design means the solution can work with any datastore.
 
-- __Batching__: You may specify that Operations can be batched (ex. writes) and then specify constraints, like how often Operations should be flushed, maximum batch size, datastore capacity, etc. Batcher will send you batches of Operations ready for you to process within all your constraints.
+- __Batching__: You may specify that Operations can be batched (ex. writes) and then specify constraints, like how often Operations should be flushed, maximum batch size, maximum concurrency, datastore capacity, etc. Batcher will send you batches of Operations ready for you to process within all your constraints.
```
Comment (KeithJRome) [inline README.md]: By "maximum concurrency", is this referring to concurrent _batches_? If so, then it could be more clear to use the "inflight" terminology you introduced above so that the reader is sure that this is what you mean here.

Or does this refer to the number of operations allowed across all batches (assuming that each batch might be fanning out their operations asynchronously)? In that case, how does this differ from "maximum batch size"?

In other words, is this referring to overall concurrency at the batch level, or the concurrency of operations within and across those batches?

File: batcher.go [hunk 4]
```diff
@@ -425,35 +484,57 @@ func (r *Batcher) Start() (err error) {
...
+				// define the func for adding to a batch
+				add := func(op IOperation) bool {
+					switch {
+					case op == nil:
+						// op can be nil when the buffer is closed; no need to continue
+						return false
+					case op.IsBatchable():
+						watcher := op.Watcher()
+						batch, ok := batches[watcher]
+						if (batch == nil || !ok) && !r.tryReserveBatchSlot() {
```
Comment (awkwardindustries) [inline batcher.go]: I read this as there was an issue getting the batch, but the `!r.tryReserveBatchSlot()` is less clear to me since that implementation does a bit more than just return a flag. In other words, if `tryReserveBatchSlot` is successful, would we need to re-call `batch, ok := batches[watcher]` in an attempt to get a non-nil batch? Or is that why it's added to pending so that it can be handled later...

File: README.md [hunk 5]
```diff
@@ -193,7 +197,7 @@ batcher := gobatcher.NewBatcherWithBuffer(buffer).
...
+- __AuditInterval__ [DEFAULT: 10s]: This determines how often the Target and Inflight variables are audited to ensure they are accurate. The Target is manipulated with atomic Operations and abandoned batches are cleaned up after MaxOperationTime so Target should always be accurate. Inflight is a channel and abandoned batches are cleaned up after MaxOperationTime so it should also always be accurate. Therefore, we should expect to only see "audit-pass" and "audit-skip" events. This audit interval is a failsafe that if the buffer is empty and the MaxOperationTime (on Batcher only; Watchers are ignored) is exceeded and the Target or Inflight is greater than zero, they are reset and an "audit-fail" event is raised. Since Batcher is a long-lived process, this audit helps ensure a broken process does not monopolize SharedCapacity when it isn't needed or that batches are stalled indefinately.
```
Comment (KeithJRome) [inline README.md]: Is this used strictly for auditing, or is it also a watchdog? It sounds like a combination of both, since it prunes zombied reservations in addition to raising audit pass/skip/fail events. And since it's manipulating those reservations, but I don't think it takes any steps to pre-empt/terminate those zombies, there should probably be a statement here that indicates that those stalled batches are abandoned but **not** terminated/canceled.

File: README.md [hunk 6]
```diff
@@ -203,6 +207,8 @@ batcher := gobatcher.NewBatcherWithBuffer(buffer).
 
 - __WithEmitBatch__ [OPTIONAL]: DO NOT USE IN PRODUCTION. For unit testing it may be useful to batches that are raised across all Watchers. Setting this flag causes a "batch" event to be emitted with the operations in a batch set as the metadata (see the sample). You would not want this in production because it will diminish performance but it will also allow anyone with access to the batcher to see operations raised whether they have access to the Watcher or not.
 
+- __WithMaxConcurrentBatches__ [OPTIONAL]: If you specify this option, Batcher will ensure that the number of batches being processed at one time does not exceed this value. Batches are still only produced on the FlushInterval. When a batch is marked done, the concurrency slot is freed for another batch.
```
Comment (KeithJRome) [inline README.md]: Is this related to the "inflight" term above? If so, it should be worded that way for consistency.

File: README.md [hunk 7]
```diff
@@ -379,12 +385,12 @@ A Batcher with a rate limiter depends on each operation having a cost. The follo
...
 
 - There is currently no way to change capacity in the rate limiters once they are provisioned, but there is no identified use-case yet for this feature.
 
 - There is currently no good way to model a datastore that autoscales but might require some time to increase capacity. Ideally something that allowed for capacity to increase by "no more than x amount over y time" would helpful. This could be a rate limiter or a feature that is added to existing rate limiters.
 
 - The pause logic is a simple fixed amount of time to delay new batches, but it might be nice to have an exponential back-off.
+
+- Currently the only shared capacity rate limiter is for Azure. It would be nice to add support for Zookeeper, Consul, or etcd.
```
Comment (KeithJRome) [inline README.md]: Another option worth mentioning here might be redis/redsync

**Review by awkwardindustries — APPROVED** [review]
Looks good! I commented on a couple of places that had me reading and re-reading to see if we might be able to make it more clear, but they would not effect functionality.

File: batcher.go [hunk 8]
```diff
@@ -250,29 +270,29 @@ func (r *Batcher) Flush() {
...
 
 // This tells you how much capacity the Batcher believes it needs to process everything outstanding. Outstanding operations include those in
 // the buffer and operations and any that have been sent as a batch but not marked done yet.
 func (r *Batcher) NeedsCapacity() uint32 {
-	return r.getTarget()
+	return r.Target()
 }
 
-func (r *Batcher) getTarget() uint32 {
+func (r *Batcher) Target() uint32 {
```
Comment (KeithJRome) [inline batcher.go]: IMO "TargetCapacity" or even "DesiredCapacity" would be a less vague name for this.

File: batcher.go [hunk 9]
```diff
@@ -288,6 +308,40 @@ func (r *Batcher) incTarget(val int) {
 	} // else is val=0, do nothing
 }
 
+func (r *Batcher) tryReserveBatchSlot() bool {
+	if r.maxConcurrentBatches == 0 {
```
Comment (KeithJRome) [inline batcher.go]: This seems like it should be documented, if setting this config value to zero disables the limiter entirely.

File: batcher.go [hunk 10]
```diff
@@ -317,44 +371,48 @@ func (r *Batcher) Start() (err error) {
...
-				}
+			// the batch is "done" when the ProcessBatch func() finishes or the maxOperationTime is exceeded
+			maxOperationTime := r.maxOperationTime
+			if watcher.MaxOperationTime() > 0 {
+				maxOperationTime = watcher.MaxOperationTime()
+			}
+			select {
+			case <-waitForDone:
+			case <-time.After(maxOperationTime):
+			}
```
Comment (KeithJRome) [inline batcher.go]: I'm wondering if this would be more clear if structured differently...

if watcher.max... > 0 {
    select {
        case <-waitForDone:
        case <-time.After(watcher.max...):
    }
} else {
    <-waitForDone
}

File: batcher.go [hunk 11]
```diff
@@ -392,9 +450,16 @@ func (r *Batcher) Start() (err error) {
...
 				// ensure that if the buffer is empty and everything should have been flushed, that target is set to 0
 				if len(r.buffer) < 1 && time.Since(lastFlushWithRecords) > r.maxOperationTime {
-					if r.trySetTargetToZero() {
-						r.emit(AuditFailEvent, 0, "an audit revealed that the target should be zero but was not.", nil)
-					} else {
+					targetIsZero := r.confirmTargetIsZero()
+					inflightIsZero := r.confirmInflightIsZero()
+					switch {
+					case !targetIsZero && !inflightIsZero:
+						r.emit(AuditFailEvent, 0, AuditMsgFailureOnTargetAndInflight, nil)
```
Comment (KeithJRome) [inline batcher.go]: This feels strange to emit an event that is a combination of two other events. I feel like this should just raise both of the other events instead adding a third "combo" event.

**Review by KeithJRome — CHANGES_REQUESTED** [review]

File: batcher.go (hunk 11 above)
Comment (plasne) [reply batcher.go]: I originally had it that way; I changed it because it seemed weird that a single audit could have 1 event (success), sometimes 1 event (fail), and sometimes multiple events (fail).

File: batcher.go (hunk 2 above)
Comment (plasne) [reply batcher.go]: this code block was changed to a new implementation.

File: batcher.go (hunk 4 above)
Comment (plasne) [reply batcher.go]: Really none of those things, so maybe my wording is unclear. "tryReserveBatchSlot" is checking that if max-concurrency is implemented that the _new_ batch (i.e. one that is non-existent or nil) has a slot to execute in. In other words, if a new batch is going to be created, it needs to not violate the concurrency constraint.

File: batcher.go (hunk 10 above)
Comment (plasne) [reply batcher.go]: The behavior is actually different than you are showing. The behavior is waitForDone or wait for the maxOperationTime. But maxOperationTime can either come from Batcher (which always has one because there is a default) or the Watcher which may or may not have a maxOperationTime (but if it does, it takes precedent).

File: README.md (hunk 3 above)
Comment (plasne) [reply README.md]: I reworded it to make it more clear. It refers to Inflight (the number of batches that can be processing at any given point in time). It has nothing to do with the number of operations, only the number of batches.

File: README.md (hunk 5 above)
Comment (plasne) [reply README.md]: I added some verbiage to clarify it abandons watching batches that exceed the maximum time.

File: README.md (hunk 7 above)
Comment (plasne) [reply README.md]: Added.

File: batcher.go (hunk 8 above)
Comment (plasne) [reply batcher.go]: There was already a property called "NeedsCapacity" that returns target, so this was just a mistake adding this one. I actually even removed the private property because it wasn't used.

File: batcher.go (hunk 9 above)
Comment (plasne) [reply batcher.go]: I updated the documentation.

File: README.md (hunk 6 above)
Comment (plasne) [reply README.md]: I updated the wording.

File: v2/batcher.go [hunk 12]
```diff
@@ -288,6 +309,88 @@ func (r *Batcher) incTarget(val int) {
...
+
+		// decrement target
+		var total int = 0
+		for _, op := range batch {
+			total += int(op.Cost())
+		}
+		r.incTarget(-total)
+
+		// remove from inflight
+		r.releaseBatchSlot()
```
Comment (awkwardindustries) [inline v2/batcher.go:384]: Any cases where this might not be reached?

File: v2/errors.go [hunk 13]
```diff
@@ -42,6 +42,20 @@ func (e BufferNotAllocated) Error() string {
...
+func (e UnspecifiedOperationToRemoveFromBuffer) Error() string {
+	return "there was no operation specified to be removed from the buffer."
+}
+
+type InvalidChainOnRemoveFromBuffer struct{}
+
+func (e InvalidChainOnRemoveFromBuffer) Error() string {
+	return "the previous supplied operation was not valid when calling remove from buffer."
+}
+*/
```
Comment (awkwardindustries) [inline v2/errors.go]: Is this needed any longer?

**Review by awkwardindustries — APPROVED** [review]

File: v2/batcher.go (hunk 12 above)
Comment (plasne) [reply v2/batcher.go:384]: The select blocks until MaxOperationTime (Watcher if it has one, Batcher if not; 1 min default) or the goroutine wrapping the callback exits (even if it were a panic). So the only way this block doesn't get hit is if the programmer has chosen an effectively forever MaxOperationTime (like 100 years) and have blocked in the callback. Even if that happens though, at the AuditInterval those will be cleaned up provided the Batcher MaxOperationTime is exceeded. So again, they could defeat this by choosing a 100 year Batcher MaxOperationTime.

SUMMARY: I feel like it is as protected as it can be outside of the dev choosing to defeat it.

File: v2/errors.go (hunk 13 above)
Comment (plasne) [reply v2/errors.go]: I dropped them; this was leftover from a failed attempt at refactoring.

File: v2/batcher_test.go [hunk 14]
```diff
@@ -890,3 +947,153 @@ func TestManualFlush(t *testing.T) {
 		assert.Fail(t, "expected the manual flush to have completed the batch before the timeout")
 	}
 }
+
+type TestMaxConcurrentBatchesSuite struct {
```
Comment (KeithJRome) [inline v2/batcher_test.go:951]: Good to see test suites used for this 👍

File: v2/buffer.go [hunk 15]
```diff
@@ -0,0 +1,172 @@
+package batcher
+
+import (
+	"errors"
+	"sync"
+)
+
+type ibuffer interface {
```
Comment (KeithJRome) [inline v2/buffer.go:8]: I understand why this has the "i" prefix, but it doesn't feel right. Perhaps the interface should be "buffer" and the struct "linkedBuffer". Just a suggestion though.

**Review by KeithJRome — APPROVED** [review]
Thanks for incorporating my feedback. I've left one small suggestion but am approving the changes as-is.
